src/analysis/normed/group/add_circle.lean

```diff
@@ -0,0 +1,234 @@
+ * Measure space structure
+ * Lie group structure
+ * Exponential equivalence to `circle`
+ * The fact `inner_product_geometry.angle (real.cos θ) (real.sin θ) = ∥(θ : real.angle)∥`
+
+-/
+
+noncomputable theory
+
+open set int (hiding mem_zmultiples_iff) add_subgroup
+
+variables {𝕜 : Type*}
+
+/-- The "additive circle": `𝕜 ⧸ (ℤ ∙ p)`. See also `circle` and `real.angle`. -/
+@[derive [add_comm_group, topological_space, topological_add_group, inhabited, has_coe_t 𝕜],
+  nolint unused_arguments]
+def add_circle [linear_ordered_add_comm_group 𝕜] [topological_space 𝕜] [order_topology 𝕜] (p : 𝕜) :=
+𝕜 ⧸ zmultiples p
+
+namespace add_circle
+
+section linear_ordered_field
+
+variables [linear_ordered_field 𝕜] [topological_space 𝕜] [order_topology 𝕜] (p q : 𝕜)
+
+@[continuity] protected lemma continuous_mk' :
+  continuous (quotient_add_group.mk' (zmultiples p) : 𝕜 → add_circle p) :=
+continuous_coinduced_rng
+
+/-- An auxiliary definition used only for constructing `add_circle.equiv_add_circle`. -/
+private def equiv_add_circle_aux (hp : p ≠ 0) : add_circle p →+ add_circle q :=
+quotient_add_group.lift _
+  ((quotient_add_group.mk' (zmultiples q)).comp $ add_monoid_hom.mul_right (p⁻¹ * q))
+  (λ x h, by obtain ⟨z, rfl⟩ := mem_zmultiples_iff.1 h; simp [hp, mul_assoc (z : 𝕜), ← mul_assoc p])
+
+/-- The rescaling equivalence between additive circles with different periods. -/
+def equiv_add_circle (hp : p ≠ 0) (hq : q ≠ 0) : add_circle p ≃+ add_circle q :=
+{ to_fun := equiv_add_circle_aux p q hp,
+  inv_fun := equiv_add_circle_aux q p hq,
+  left_inv := by { rintros ⟨x⟩, show quotient_add_group.mk _ = _, congr, field_simp [hp, hq], },
+  right_inv := by { rintros ⟨x⟩, show quotient_add_group.mk _ = _, congr, field_simp [hp, hq], },
+  .. equiv_add_circle_aux p q hp }
+
+@[simp] lemma equiv_add_circle_apply_mk (hp : p ≠ 0) (hq : q ≠ 0) (x : 𝕜) :
+  equiv_add_circle p q hp hq (x : 𝕜) = (x * (p⁻¹ * q) : 𝕜) :=
+rfl
+
+@[simp] lemma equiv_add_circle_symm_apply_mk (hp : p ≠ 0) (hq : q ≠ 0) (x : 𝕜) :
+  (equiv_add_circle p q hp hq).symm (x : 𝕜) = (x * (q⁻¹ * p) : 𝕜) :=
+rfl
+
+variables [floor_ring 𝕜]
+
+/-- The natural equivalence between `add_circle p` and the half-open interval `[0, p)`. -/
+def equiv_Ico (hp : 0 < p) : add_circle p ≃ Ico 0 p :=
+{ inv_fun := quotient_add_group.mk' _ ∘ coe,
+  to_fun :=
+  begin
+    have h₀ : ∀ (x : 𝕜), 0 ≤ fract (x / p) * p := λ x, mul_nonneg (fract_nonneg _) hp.le,
+    have h₁ : ∀ (x : 𝕜), fract (x / p) * p < p := λ x, mul_lt_of_lt_one_left hp $ fract_lt_one _,
+    letI := quotient_add_group.left_rel (zmultiples p),
+    have : ∀ (x y : 𝕜) (h: x ≈ y),
+      (⟨fract (x / p) * p, h₀ x, h₁ x⟩ : Ico 0 p) = (⟨fract (y / p) * p, h₀ y, h₁ y⟩ : Ico 0 p),
+    { intros,
+      obtain ⟨z, hz⟩ := mem_zmultiples_iff.mp (quotient_add_group.left_rel_apply.mp h),
+      ext,
+      simp [(by { rw hz, abel, } : y = z • p + x), mul_div_cancel _ hp.ne.symm, add_div], },
+    exact quotient.lift _ this,
+  end,
```

Comment:
This feels like it's much more complicated than necessary (compared to the definition of `real.angle.to_real`). I'd expect something like `λ x, ⟨(to_Ico_mod_periodic 0 hp).lift x, to_Ico_mod_mem_Ico 0 hp x⟩` to suffice (and maybe some other proofs might also be simpler, considering those for `to_real`).

Reply:
(Maybe `to_Ico_mod_mem_Ico` doesn't have quite the right type, because it actually asserts membership of `Ico 0 (0 + p)`, but correcting for that should be straightforward.)

Reply:
Thanks, yes this is just a silly proof I bashed out so I could push something sorry-free.

I hope to return to this PR in a day or two and I expect using `to_Ico_mod_mem_Ico` for inspiration here will help. Thanks again.